service/src/main/java/bio/terra/cbas/util/BackfillOriginalWorkspaceIdService.java

```diff
@@ -0,0 +1,175 @@
+package bio.terra.cbas.util;
+
+import bio.terra.cbas.config.CbasContextConfiguration;
+import bio.terra.cbas.dao.MethodDao;
+import bio.terra.cbas.dao.MethodVersionDao;
+import bio.terra.cbas.dao.RunSetDao;
+import bio.terra.cbas.models.Method;
+import bio.terra.cbas.models.MethodVersion;
+import bio.terra.cbas.models.RunSet;
+import bio.terra.common.db.WriteTransaction;
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+/*
+NOTE: This class is inherently temporary.  Once all original workspace IDs have been backfilled,
+      this entire class should be deleted, along with its method calls in BackfillOriginalWorkspaceIdBean.java
+ */
+@Service
+public class BackfillOriginalWorkspaceIdService {
+  private final Logger logger = LoggerFactory.getLogger(BackfillOriginalWorkspaceIdService.class);
+  @Autowired CbasContextConfiguration cbasContextConfig;
+  @Autowired RunSetDao runSetDao;
+  @Autowired MethodDao methodDao;
+  @Autowired MethodVersionDao methodVersionDao;
+
+  public BackfillOriginalWorkspaceIdService(
+      CbasContextConfiguration cbasContextConfig,
+      RunSetDao runSetDao,
+      MethodDao methodDao,
+      MethodVersionDao methodVersionDao) {
+    this.cbasContextConfig = cbasContextConfig;
+    this.methodDao = methodDao;
+    this.runSetDao = runSetDao;
+    this.methodVersionDao = methodVersionDao;
+  }
+
+  @WriteTransaction
+  public void backfillAll() {
+    if (cbasContextConfig
+        .getWorkspaceCreatedDate()
+        .isEqual(OffsetDateTime.parse("0000-01-01T00:00:00.000000Z"))) {
+      logger.info(
+          "Skipping BackfillOriginalWorkspaceId operation. This could indicate that workspaceCreatedDate in application.yml has not been configured properly.");
+    } else {
+      try {
+        backfillMethods();
+        backfillRunSets();
+        backfillMethodVersions();
+      } catch (Exception e) {
+        // we want CBAS to fail if this operation fails, just like a liquibase migration
+        throw new RuntimeException(
+            "There was a problem while trying to backfill original workspace IDs: %s".formatted(e));
+      }
+    }
+  }
+
+  @WriteTransaction
```

Comment:
This is kind of a giant transaction... can we make it any more specific to things that really need to happen atomically?

Reply:
(eg, each runset is updated in its own transaction?)

Reply:
yeah, thanks for pointing this out - my thinking was: "we want all this to succeed or fail together, so put it in a transaction". but I see your point completely

Reply:
looking at this more closely - each of these methods (e.g. for run_set, method, method_version...) only makes one update per object type per iteration (e.g. it loops over a list of run sets, and calls one DAO method)

that makes me think that the WriteTransaction annotation is not really needed here at all.

hopefully I'm not misunderstanding something important